src/main/kotlin/org/rust/ide/formatter/RustBlock.kt

```diff
@@ -21,6 +26,23 @@ class RustBlock(
     }
 
     override fun getSpacing(child1: Block?, child2: Block): Spacing? {
+        if (child1 is AbstractRustBlock && child2 is AbstractRustBlock) {
+            val elt1 = child1.node.elementType
+            val elt2 = child2.node.elementType
+            val psi1 = child1.node.psi
+            val psi2 = child2.node.psi
+            val parent = psi1.parent
+
+            if (psi1 is RustOuterAttr && (psi2 is RustOuterAttr || parent is RustItem)) {
+                return RustSpacing.lineBreak()
+            }
+
+            if (psi1.isTopLevelItem && psi2.isTopLevelItem) {
+                return RustSpacing.lineBreak(
+                    keepLineBreaks = ctx.commonSettings.KEEP_LINE_BREAKS,
+                    keepBlankLines = ctx.commonSettings.KEEP_BLANK_LINES_IN_DECLARATIONS)
+            }
+        }
         return ctx.spacingBuilder.getSpacing(this, child1, child2)
     }
```

Comment:
I'm currently writing this the "classic" way - how it's implemented in Python, Go, Java, etc etc formatters. Though Kotlin formatter does this differently: it [greatly extends SpacingBuilder](https://github.com/JetBrains/kotlin/blob/master/idea/formatter/src/org/jetbrains/kotlin/idea/formatter/KotlinSpacingBuilder.kt) to empower creating complex spacing rules with a declarative way.